students/stefan_lund/Lesson_5/Lesson_5_mailroom/mailroom_5_functions.py

```diff
@@ -0,0 +1,323 @@
+    # create letters
+    for name, donations in sorted(data.items()):
+        donation = 0
+        for date in donations:
+            donation += donations[date]
+
+        total = float(donation)
+
+        letter = letter_template.format(name=name,
+                                        amount=total,
+                                        date=month+"/"+day+"/"+year)
+        donor = "_".join(name.split())
+        letter_file = donor + "__" + year + "_" + month + "_" + day
+        letters.append((letter, letter_file))
+
+
+    # print letter to screen and store letter
+
+    # create the directory to store the letters if it doesn't exist
+    pth = os.path.join(os.getcwd(), "letter_to_everyone")
+    if not os.path.exists(pth):
+        os.makedirs(pth)
+
+    for letter, letterfile in letters:
+        print("\n", letter)
+        destination = os.path.join(pth, letterfile)
+
+        try:
+            with open(destination, 'w') as outfile:
+                for line in letter:
+                    outfile.write(line)
+        except FileNotFoundError:
+            print("Could not find the file")
+
+    print("stored letters in ", pth)
+
+    return "start"
+
+
+# "thank you" menu functions
+
+def main_menu():
+    """
+    sole function is to switch to the "start" menue
+    """
+    return "start"
+
+def list_of_names():
+    """
+        prints out a list of names currently in the "data" dictionary
+    """
+    data = mailroom_5_read_write_data.data
+
+    # determine width of fields to print in
+    name_list = data.keys()
+    name_length = []
+    # list comprehension
+    [name_length.append(len(name)) for name in name_list]
+    name_len = max(name_length) + 3
+
+    # print table header
+    temp_str = "Donor Name"
+    temp_form = "\n| {:<{nl}}|"
+    header = temp_form.format(temp_str, nl=name_len)
+    frame = "\n+" + "-" * (name_len + 1) + "+"
+    print(frame, header, frame)
+
+    # print each name in alphabetical order
+    for name in sorted(name_list):
+        temp_form = "| {:<{nl}}|"
+        line = temp_form.format(name, nl=name_len)
+        print(line)
+
+    return "thank_you"
+
+
+def enter_name():
+    """
+    updates data dictionary with the donation from a new or a present donor
+    and prints out a thank you letter
+    and stores the letter in cwd folder "thank_you_letter"
+    """
+    data = mailroom_5_read_write_data.data
+    today = todays_date()
+    year, month, day = today[0], today[1], today[2]  #  str type
+
+    # user enter name, check if input is letters or space
+    name_of_letters = False
+    while not name_of_letters:
+        name = input("\nEnter name, 'First Last': ")
+        name_of_letters = is_letters_or_space(name)
+
+    # user enter new_amount, check if new_amount is a number, 2 decimals or less
+    digits_in_amount = False
+    while not digits_in_amount:
+        new_amount = input("\nEnter amount donated: ")
+        digits_in_amount = is_digit_or_period(new_amount)

```

Comment:
this is where you should replace the code with exceptions handling, which is really the right approach. Instead of writing complicated checking functions - try the user input:
```
while True:
    try:
        amount = float(input("..."))
    except ValueError:
        print("invalid amount value, try again")
    else:
        break
```